pkg/insight/provider.go

```diff
@@ -0,0 +1,193 @@
+}
+
+type provider struct {
+	store Store
+}
+
+func NewProvider(s Store) Provider {
+	return &provider{
+		store: s,
+	}
+}
+
+// TODO: Add cache layer.
+func (p *provider) GetApplicationCounts(ctx context.Context, projectID string) (*ApplicationCounts, error) {
+	data, err := p.store.GetApplications(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	counts := buildApplicationCounts(data)
+	return &counts, nil
+}
+
+// TODO: Add cache layer.
+func (p *provider) GetDeploymentFrequencyDataPoints(ctx context.Context, projectID, appID string, labels map[string]string, rangeFrom, rangeTo int64) ([]*model.InsightDataPoint, error) {
+	ds, err := p.store.ListCompletedDeployments(ctx, projectID, rangeFrom, rangeTo)
+	if err != nil {
+		return nil, err
+	}
+
+	points := buildDeploymentFrequencyDataPoints(ds, appID, labels)
+	return points, nil
+}
+
+// TODO: Add cache layer.
+func (p *provider) GetDeploymentChangeFailureRateDataPoints(ctx context.Context, projectID, appID string, labels map[string]string, rangeFrom, rangeTo int64) ([]*model.InsightDataPoint, error) {
+	ds, err := p.store.ListCompletedDeployments(ctx, projectID, rangeFrom, rangeTo)
+	if err != nil {
+		return nil, err
+	}
+
+	points := buildDeploymentChangeFailureRateDataPoints(ds, appID, labels)
+	return points, nil
+}
+
+func buildDeploymentFrequencyDataPoints(ds []*DeploymentData, appID string, labels map[string]string) []*model.InsightDataPoint {
+	ds = filterDeploymentData(ds, appID, labels)
+	if len(ds) == 0 {
+		return []*model.InsightDataPoint{}
+	}
+
+	var (
+		out      = make([]*model.InsightDataPoint, 0)
+		curPoint *model.InsightDataPoint
+	)
+	for _, d := range ds {
+		if curPoint == nil || curPoint.Timestamp != d.CompletedAtDay {
+			curPoint = &model.InsightDataPoint{
+				Timestamp: d.CompletedAtDay,
+			}
+			out = append(out, curPoint)
+		}
+		curPoint.Value += 1
+	}
+
+	return out
+}
+
+func buildDeploymentChangeFailureRateDataPoints(ds []*DeploymentData, appID string, labels map[string]string) []*model.InsightDataPoint {
+	ds = filterDeploymentData(ds, appID, labels)
+	if len(ds) == 0 {
+		return []*model.InsightDataPoint{}
+	}
+
+	var (
+		out                              = make([]*model.InsightDataPoint, 0)
+		curPoint *model.InsightDataPoint = nil
+		curTotal                         = 0
+		curFails                         = 0
+	)
+	for _, d := range ds {
+		if curPoint == nil || curPoint.Timestamp != d.CompletedAtDay {
+			if curPoint != nil {
+				curPoint.Value = float32(curFails) / float32(curTotal)
+				curTotal = 0
+				curFails = 0
+			}
+			curPoint = &model.InsightDataPoint{
+				Timestamp: d.CompletedAtDay,
+			}
+			out = append(out, curPoint)
+		}
+		curTotal += 1
+		if d.CompleteStatus == model.DeploymentStatus_DEPLOYMENT_FAILURE.String() {
+			curFails += 1
+		}
+	}
+	if curPoint != nil {
+		curPoint.Value = float32(curFails) / float32(curTotal)
+	}
+
```

Comment:
this part blows my mind and looks unsafe since elements in out are all shared the same pointer curPoint 🤔

Reply:
@khanhtc1202 I don't think so. A new `curPoint` will be created and added whenever the `completedAtDay` changes.
And here https://github.com/pipe-cd/pipecd/pull/3993/files#diff-0a0f140ec136a5b33865083d31eaf5bc1f44f9de4b190bdb50e7f2883dcccbe8R77 is its test.